File: frontend/lib/api/job-types.ts
import apiClient from "./client";

export interface WorkflowProfile {
  id: number;
  code: string;
  name: string;
  description?: string;
  is_active?: boolean;
  is_predefined?: boolean;
  sort_order?: number;
  skip_inspection: boolean;
  skip_diagnosis: boolean;
  skip_customer_approval: boolean;
  skip_quality_check: boolean;
  auto_approve_on_create: boolean;
  apply_service_bundle_on_create: boolean;
  allows_fast_track_to_approved: boolean;
}

export interface JobTypeWriteInput {
  code?: string;
  name: string;
  category: string;
  description?: string;
  workflow_profile: number;
  is_active?: boolean;
  sort_order?: number;
  requires_inspection?: boolean;
  requires_diagnosis?: boolean;
  requires_approval?: boolean;
  quality_check_required?: boolean;
  allows_bundle?: boolean;
  sets_warranty_flag?: boolean;
  sets_insurance_flag?: boolean;
  default_revenue_product?: number | null;
  default_service_fee?: string | null;
}

export interface JobType {
  id: number;
  code: string;
  name: string;
  category: string;
  category_display?: string;
  description?: string;
  workflow_profile: WorkflowProfile;
  is_active: boolean;
  sort_order: number;
  requires_inspection: boolean;
  requires_diagnosis: boolean;
  requires_approval: boolean;
  quality_check_required: boolean;
  allows_bundle: boolean;
  sets_warranty_flag: boolean;
  sets_insurance_flag: boolean;
  default_revenue_product?: number | null;
  default_revenue_product_name?: string | null;
  default_revenue_product_code?: string | null;
  default_service_fee?: string | null;
}

export interface JobTypeListResponse {
  count: number;
  next: string | null;
  previous: string | null;
  results: JobType[];
}

export const jobTypesApi = {
  list: async (params?: {
    category?: string;
    allows_bundle?: boolean;
    active_only?: boolean;
    workflow_profile?: number;
  }): Promise<JobTypeListResponse> => {
    const response = await apiClient.get("/workorders/job-types/", { params });
    return response.data;
  },

  get: async (code: string): Promise<JobType> => {
    const response = await apiClient.get(`/workorders/job-types/${code}/`);
    return response.data;
  },

  create: async (data: JobTypeWriteInput): Promise<JobType> => {
    const response = await apiClient.post("/workorders/job-types/", data);
    return response.data;
  },

  update: async (code: string, data: Partial<JobTypeWriteInput>): Promise<JobType> => {
    const response = await apiClient.patch(`/workorders/job-types/${code}/`, data);
    return response.data;
  },

  remove: async (code: string): Promise<void> => {
    await apiClient.delete(`/workorders/job-types/${code}/`);
  },
};

export const workflowProfilesApi = {
  list: async (): Promise<WorkflowProfile[]> => {
    const response = await apiClient.get("/workorders/workflow-profiles/");
    const data = response.data;
    return Array.isArray(data) ? data : data.results ?? [];
  },
};

export function isFastTrackJobType(jobType?: JobType | null): boolean {
  return Boolean(jobType?.workflow_profile?.allows_fast_track_to_approved);
}

export function jobTypeRequiresBundle(jobType?: JobType | null): boolean {
  return Boolean(
    jobType?.allows_bundle && jobType?.workflow_profile?.apply_service_bundle_on_create
  );
}

export function getJobTypeDisplayName(
  workOrder?: {
    job_type_detail?: JobType | null;
    job_type?: number | JobType | null;
    maintenance_type?: string | null;
  } | null
): string {
  if (!workOrder) return "General repair";
  if (workOrder.job_type_detail?.name) return workOrder.job_type_detail.name;
  if (typeof workOrder.job_type === "object" && workOrder.job_type?.name) {
    return workOrder.job_type.name;
  }
  if (workOrder.maintenance_type === "routine") return "Routine service";
  return "General repair";
}

File: apps/workorders/job_types.py
"""
Configurable work order job types and workflow profiles.

Job types describe *what* work is being done (Brake Service, Warranty Repair, etc.).
Workflow profiles describe *how* the work order flows through the shop.
"""

from decimal import Decimal

from django.core.validators import MinValueValidator
from django.db import models
from django.utils.text import slugify


class WorkflowProfile(models.Model):
    """Template controlling which work-order stages apply and default skip rules."""

    code = models.SlugField(max_length=50, unique=True)
    name = models.CharField(max_length=100)
    description = models.TextField(blank=True)
    is_active = models.BooleanField(default=True)
    is_predefined = models.BooleanField(default=False)
    sort_order = models.PositiveIntegerField(default=0)

    skip_inspection = models.BooleanField(default=False)
    skip_diagnosis = models.BooleanField(default=False)
    skip_customer_approval = models.BooleanField(default=False)
    skip_quality_check = models.BooleanField(default=False)
    auto_approve_on_create = models.BooleanField(default=False)
    apply_service_bundle_on_create = models.BooleanField(default=False)
    allows_fast_track_to_approved = models.BooleanField(
        default=False,
        help_text='Allow draft → approved when bundle/tasks are ready (routine fast-track).',
    )

    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)

    class Meta:
        ordering = ['sort_order', 'name']
        verbose_name = 'Workflow Profile'
        verbose_name_plural = 'Workflow Profiles'

    def __str__(self):
        return self.name


class JobType(models.Model):
    """Admin-manageable catalog of work order job types."""

    CATEGORY_CHOICES = [
        ('repair', 'Repair'),
        ('maintenance', 'Maintenance'),
        ('diagnostic', 'Diagnostic'),
        ('inspection', 'Inspection'),
        ('body', 'Body & Paint'),
        ('commercial', 'Warranty / Insurance'),
        ('installation', 'Installation'),
    ]

    code = models.SlugField(max_length=50, unique=True)
    name = models.CharField(max_length=100)
    category = models.CharField(max_length=20, choices=CATEGORY_CHOICES, default='repair')
    description = models.TextField(blank=True)
    workflow_profile = models.ForeignKey(
        WorkflowProfile,
        on_delete=models.PROTECT,
        related_name='job_types',
    )
    is_active = models.BooleanField(default=True)
    is_predefined = models.BooleanField(default=False)
    sort_order = models.PositiveIntegerField(default=0)

    requires_inspection = models.BooleanField(default=True)
    requires_diagnosis = models.BooleanField(default=True)
    requires_approval = models.BooleanField(default=True)
    quality_check_required = models.BooleanField(default=True)
    allows_bundle = models.BooleanField(
        default=False,
        help_text='Whether a service bundle may be selected at check-in.',
    )

    default_service_type = models.ForeignKey(
        'vehicles.ServiceType',
        on_delete=models.SET_NULL,
        null=True,
        blank=True,
        related_name='default_job_types',
    )
    default_service_bundle = models.ForeignKey(
        'inventory.ServiceBundle',
        on_delete=models.SET_NULL,
        null=True,
        blank=True,
        related_name='default_job_types',
    )

    sets_warranty_flag = models.BooleanField(default=False)
    sets_insurance_flag = models.BooleanField(default=False)

    default_revenue_product = models.ForeignKey(
        'accounting.RevenueProduct',
        on_delete=models.SET_NULL,
        null=True,
        blank=True,
        related_name='default_job_types',
        help_text='Default income category when invoicing this job type with no tasks/parts.',
    )
    default_service_fee = models.DecimalField(
        max_digits=10,
        decimal_places=2,
        null=True,
        blank=True,
        validators=[MinValueValidator(Decimal('0'))],
        help_text='Optional flat fee override for this job type (otherwise uses income category default price).',
    )

    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)

    class Meta:
        ordering = ['sort_order', 'name']
        verbose_name = 'Job Type'
        verbose_name_plural = 'Job Types'
        indexes = [
            models.Index(fields=['is_active', 'category']),
            models.Index(fields=['code']),
        ]

    def __str__(self):
        return self.name

    def save(self, *args, **kwargs):
        if not self.code and self.name:
            base = slugify(self.name) or 'job-type'
            code = base
            suffix = 2
            while JobType.objects.filter(code=code).exclude(pk=self.pk).exists():
                code = f'{base}-{suffix}'
                suffix += 1
            self.code = code
        super().save(*args, **kwargs)

    def apply_defaults_to_work_order(self, work_order, *, overwrite=False):
        """Apply job-type default flags onto a work order (typically at create)."""
        fields_to_update = []

        def _set(field_name, value):
            if overwrite or getattr(work_order, field_name) in (None, ''):
                setattr(work_order, field_name, value)
                fields_to_update.append(field_name)

        _set('requires_approval', self.requires_approval)
        _set('quality_check_required', self.quality_check_required)

        if self.sets_warranty_flag:
            work_order.is_warranty = True
            fields_to_update.append('is_warranty')
        if self.sets_insurance_flag:
            work_order.is_insurance_claim = True
            fields_to_update.append('is_insurance_claim')

        if self.default_service_type_id and (overwrite or not work_order.service_type_id):
            work_order.service_type = self.default_service_type
            fields_to_update.append('service_type')
        if self.default_service_bundle_id and (overwrite or not work_order.service_bundle_id):
            work_order.service_bundle = self.default_service_bundle
            fields_to_update.append('service_bundle')

        return fields_to_update
